test/functional/getblockstats.py

```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017-2017 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#
+# Test getblockstats rpc call
+#
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    assert_raises_rpc_error,
+)
+
+def assert_contains(data, values, check_cointains=True):
+    for val in values:
+        if (check_cointains):
+            assert(val in data)
+        else:
+            assert(val not in data)
+
+class GetblockstatsTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 2
+        self.extra_args = [['-txindex'], ['-paytxfee=0.003']]
+        self.setup_clean_chain = True
+
+    def run_test(self):
+        node = self.nodes[0]
+        node.generate(101)
+
+        node.sendtoaddress(address=self.nodes[1].getnewaddress(), amount=10, subtractfeefromamount=True)
+        node.generate(1)
+        self.sync_all()
+
+        node.sendtoaddress(address=node.getnewaddress(), amount=10, subtractfeefromamount=True)
+        node.sendtoaddress(address=node.getnewaddress(), amount=10, subtractfeefromamount=True)
+        self.nodes[1].sendtoaddress(address=node.getnewaddress(), amount=1, subtractfeefromamount=True)
+        self.sync_all()
+        node.generate(1)
+
+        start_height = 101
+        max_stat_pos = 2
+        stats = node.getblockstats(start=start_height, end=start_height + max_stat_pos)
+
+        all_values = [
+            "height",
+            "time",
+            "mediantime",
+            "txs",
+            "swtxs",
+            "ins",
+            "outs",
+            "subsidy",
+            "totalfee",
+            "utxo_increase",
+            "utxo_size_inc",
+            "total_size",
+            "total_weight",
+            "swtotal_size",
+            "swtotal_weight",
+            "total_out",
+            "minfee",
+            "maxfee",
+            "medianfee",
+            "avgfee",
+            "minfeerate",
+            "maxfeerate",
+            "medianfeerate",
+            "avgfeerate",
+            "mintxsize",
+            "maxtxsize",
+            "mediantxsize",
+            "avgtxsize",
+        ]
+        assert_contains(stats, all_values)
+        # Make sure all valid statistics are included
+        assert_contains(all_values, stats.keys())
+
+        assert_equal(stats['height'][0], start_height)
+        assert_equal(stats['height'][max_stat_pos], start_height + max_stat_pos)
+
+        assert_equal(stats['txs'][0], 1)
+        assert_equal(stats['swtxs'][0], 0)
```

Comment:
None of the transactions have witness data, so this isn't actually checking whether swtxs, swtotal_size, swtotal_weight ever output anything other than 0, or whether weight is ever anything other than size*4.

Reply:
This is true. When I wrote this segwit wasn't activated, but now it would be nice to add segwit txs to the test because otherwise this functionality isn't really being tested. Good call.